enhancements/kube-apiserver/stale-clusteroperators-detection.md

```diff
@@ -0,0 +1,87 @@
+---
+title: detection-and-reporting-of-stale-clusteroperators
+authors:
+  - "@mfojtik"
+reviewers:
+  - "@deads2k"
+approvers:
+  - "@deads2k"
+creation-date: 2022-01-31
+last-updated: 2022-01-31
+---
+
+# Detection and reporting of stale cluster operators
+
+## Summary
+
+In OpenShift 4, every active operator included in the core payload (cluster operator) updates its conditions in "
+clusteroperator" resource. Conditions determine whether specific core components are Degraded or Available or if they
+are making progress or if their state causes the cluster to be not upgradeable. These conditions are essential for
+cluster administrators to determine the health of their clusters, and they are usually the first thing Red Hat support
+is going to look at in case there is a problem with the cluster.
+
+Unfortunately, in some edge cases, these conditions might not reflect the current state of the cluster. For example,
+since they depend on the running operator, these conditions might become stale and outdated if the operator is not
+running or cannot contact the API server. And while the API server might still work, the result
+of `oc get clusteroperators` does not represent the cluster's current state.
+
+Therefore, we need a detection mechanism to help detect in-active operators or operators that are not actively updating
+the conditions.
+
+## Motivation
+
+In the past, there were multiple cases when either kubelet or CRI-O failed to run pods, either by KCM misfunctioning or
+some other system error. However, the static pods succeeded in running, making the Kubernetes API server available and
+accessible with `oc`. When support or engineers ran `oc get clusteroperators` against this cluster, they saw all
+operators reporting available, non-degraded and upgradeable status, but the truth was that none of the operators that
+provided these conditions were running.
+
+### Goals
+
+* Detect in-active cluster operators
+* Provide more accurate condition messages and status
+
+### Non-Goals
+
+* Make a new operator that detects other operators status
+* Create a heart-beat mechanism
+
+## Proposal
+
+To implement stale operator detection, we need a controller process that will run close to the Kubernetes API server,
+preferably using a local connection and client-based certificate for authentication to avoid dependencies on working
+auth and network when the cluster is unstable.
+
+The controller will watch the clusteroperator resource and check the `status.[]conditions.lastTransitionTime` field for
+Degraded, Available, Progressing, and Upgradeable conditions. Suppose the "lastTransitionTime" field is not updated for
```

Comment:
it's not clearly stated here, so to avoid any implementation confusion:  I would expect that this behavior would only be triggered if at least one of the above conditions is "stale".  As long as any one condition was updated in the last 10 mins, then no write is needed.  (vs treating each condition independently)

Reply:
good catch